crates/sui-types/src/temporary_store.rs

```diff
@@ -567,27 +567,34 @@ impl<'backing> TemporaryStore<'backing> {
         Ok(gas[0])
     }
 
-    pub fn delete_object(&mut self, id: &ObjectID, kind: DeleteKindWithOldVersion) {
+    pub fn delete_object(
+        &mut self,
+        id: &ObjectID,
+        kind: DeleteKindWithOldVersion,
+    ) -> Result<(), ExecutionError> {
         // there should be no deletion after write
         debug_assert!(self.written.get(id).is_none());
 
-        // TODO: promote this to an on-in-prod check that raises an invariant_violation
-        // Check that we are not deleting an immutable object
-        #[cfg(debug_assertions)]
+        // Check that we are not deleting an immutable object or wrapping a shared object
         if let Some(object) = self.read_object(id) {
             if object.is_immutable() {
                 // This is an internal invariant violation. Move only allows us to
                 // mutate objects if they are &mut so they cannot be read-only.
                 // In addition, gas objects should never be immutable, so gas smashing
                 // should not allow us to delete immutable objects
                 let digest = self.tx_digest;
-                panic!("Internal invariant violation in tx {digest}: Deleting immutable object {id}, delete kind {kind:?}")
+                return Err(ExecutionError::invariant_violation(format!(
+                    "Internal invariant violation in tx {}: Deleting immutable object {}",
+                    digest, id
+                )));
```

Comment:
There is a macro -- `invariant_violation!` -- that combines:

- `panic!` in debug mode.
- return an `Err(ExecutionError::invariant_violation(...))` in production.
- accepts a format string by default

So that may be more convenient to use here.

Reply:
I originally used that one, but then in the transaction tests that macro will throw the panic, making it harder to test. Then I considered to made a new macro that doesn't panic, I am happy to add that one if if will be useful to others too

Reply:
Ah okay, interesting -- the macro is actually intended to improve debuggability in those scenarios by offering a backtrace at the point of creation, rather than just the error message, because invariant violations are not typically recoverable.  If that isn't helpful, it's probably not worth creating a non-panicking macro, but I'm curious to learn more about why it made testing more difficult, because it's supposed to have the opposite effect!